Update monitor to pass basic tests

A5rocks opened the pull request:
Supersedes https://github.com/python-trio/trio-monitor/pull/3

The code kinda sucks but I didn't do obvious simplifications. Also I added a test to ensure we support the things we say we do.

A5rocks committed:
Bring trio-monitor into the future

Co-authored-by: Zach Thompson <zach@allotropic.com>

A5rocks committed:
Test against the oldest possible Trio

A5rocks committed:
Clean up a couple missed lines

Fuyukai commented on .github/workflows/ci.yml:
```diff
@@ -12,7 +12,7 @@ jobs:
       fail-fast: false
       matrix:
-        python: ["3.9", "3.10", "3.11", "3.12", "3.13", "pypy-3.10"]
+        python: ["3.8", "3.9", "3.10", "3.11", "3.12", "3.13", "pypy-3.10"]
```
why? 3.8 is very dead. really even supporting 3.9 is a bit silly but that's for another date.

A5rocks replied:
For some reason I thought 3.8 was still supported by CPython. I'll drop it.

(in defense of supporting many versions, it would be awful to have to upgrade a broken application in order to use a debug tool)

Fuyukai commented on noxfile.py:
```diff
@@ -18,6 +18,38 @@ def precommit(session):
+        "3.11": "trio>=0.21.0",
+        "3.10": "trio>=0.20.0",
+        "3.9": "trio>=0.19.0",
+        "3.8": "trio",
```
see above

Fuyukai commented on src/trio_monitor/monitor.py:
```diff
@@ -58,7 +54,7 @@ def __init__(self):
         # semi-arbitrary size, because otherwise we'll be dropping events
         # no clue how to make this better, alas.
-        self._monitoring_queue = Queue(capacity=100)
+        self._rx, self._tx = open_memory_channel(math.inf)
```
don't really like using ``math.inf`` here, i specifically did only use a small-ish queue size to prevent it growing entirely unbounded.

Fuyukai commented on src/trio_monitor/_tests/test_monitor.py:
```diff
@@ -0,0 +1,70 @@
+    run_with_monitor(main, trio_monitor.Monitor())
+
+
+def test_it_can_start_a_server():
```
I think this is the wrong way around it imo. Testing the user-focused cli output is probably not a good idea; there should be some underlying protocol that is more easily tested and then the evilness of turning the RPC into user-focused output can be entirely isolated.

A5rocks replied:
Agreed, except that (for now) the only exposed output is this console...

(One level lower would be calling methods on Monitor which is a decent idea so maybe I should do that.)

A5rocks replied:
I tried out calling methods on `Monitor` and it's a bit... not great. But at least I made a test for the core functionality of monitoring a program.

A5rocks committed:
Drop 3.8, add backpressure, update old Queue usage

A5rocks committed:
For now, don't test any commands

A5rocks committed:
Add a test for actual monitoring capabilities

Fuyukai commented on src/trio_monitor/_tests/test_monitor.py:
```diff
@@ -0,0 +1,96 @@
+            await trio.lowlevel.checkpoint()
+            await trio.lowlevel.checkpoint()
+            await trio.lowlevel.checkpoint()
+            # ^ this is necessary because monitor breaks
```
Is this an inherent problem with the monitor design or just a bug that can be fixed? I'm okay with merging as-is right now to get *something* testable but if it's a bug it should be fixed asap after this.

A5rocks replied:
I investigated and it just seemed to be an infinite loop around waking and sleeping `do_monitor`. I fixed that. (for some reason I had this code before and it didn't work, probably because I had another task that was writing out the monitor output?)

Fuyukai reviewed (approved):
I'm a bit apprehensive over the entire design of the monitor like six years later (more on this if you want to ask) but if you're going to be maintaining it I don't see anything wrong here.

A5rocks committed:
Ensure there is not an infinite loop

Previously:
 - do_monitor would get an event, waking up
 - do_monitor would sleep
 - the monitor would notice (`after_task_step`) and add a new event
 - do_monitor would wake up
 - repeat

A5rocks commented:
> I'm a bit apprehensive over the entire design of the monitor like six years later (more on this if you want to ask) but if you're going to be maintaining it I don't see anything wrong here.

I agree the design isn't great and I'm not sure how to solve it. I mean, I know some basic niceities we can do (use an actual RPC, some basic code simplifications) but I'm not so sure about the actual lifecycle of the monitor. The best I can think of is some sort of:
```py
async with start_monitor() as monitor:
  ...
```
but that's just not that satisfying a design to me...

---

If there's other parts (what other parts!) that seem poorly designed then I'd like to know also.

Fuyukai commented:
> If there's other parts (what other parts!) that seem poorly designed then I'd like to know also.

It should run in a separate thread/separate loop instance, it shouldn't be integrated into the main loop thread. I'm 90% sure before I extracted this into its own project that was the initial design, even.

A5rocks commented:
Yeah curio's monitor uses that design. I think the main point against it is that then users can't eg have a TLS stream... but actually they totally could just proxy things huh.
